src/api/schema/events/mod.rs

```diff
@@ -0,0 +1,151 @@
+/// Encoding format for the event
+pub enum EventEncodingType {
+    Json,
+    Yaml,
+}
+
+#[derive(Debug, Default)]
+pub struct EventsSubscription;
+
+#[Subscription]
+impl EventsSubscription {
+    /// A stream of events emitted from matched component(s)
+    pub async fn output_events<'a>(
+        &'a self,
+        ctx: &'a Context<'a>,
+        component_names: Vec<String>,
+        #[graphql(default = 500)] interval: u32,
+        #[graphql(default = 100, validator(IntRange(min = "1", max = "10_000")))] limit: u32,
+    ) -> impl Stream<Item = Vec<OutputEventsPayload>> + 'a {
+        let watch_rx = ctx.data_unchecked::<WatchRx>().clone();
+
+        // Client input is confined to `u32` to provide sensible bounds.
+        create_events_stream(watch_rx, component_names, interval as u64, limit as usize)
+    }
+}
+
+/// Creates an events stream based on component names, and a provided interval. Will emit
+/// control messages that bubble up the application if the sink goes away. The stream contains
+/// all matching events; filtering should be done at the caller level.
+fn create_events_stream(
+    watch_rx: WatchRx,
+    component_names: Vec<String>,
+    interval: u64,
+    limit: usize,
+) -> impl Stream<Item = Vec<OutputEventsPayload>> {
+    // Channel for receiving individual tap payloads. Since we can process at most `limit` per
+    // interval, this is capped to the same value.
+    let (tap_tx, mut tap_rx) = mpsc::channel(limit);
+
+    // The resulting vector of `Event` sent to the client. Only one result set will be streamed
+    // back to the client at a time. This value is set higher than `1` to prevent blocking the event
+    // pipeline on slower client connections, but low enough to apply a modest cap on mem usage.
+    let (mut event_tx, event_rx) = mpsc::channel::<Vec<OutputEventsPayload>>(10);
+
+    tokio::spawn(async move {
+        // Create a tap sink. When this drops out of scope, clean up will be performed on the
+        // event handlers and topology observation that the tap sink provides.
+        let _tap_sink = TapSink::new(watch_rx, tap_tx, &component_names);
+
+        // A tick interval to represent when to 'cut' the results back to the client.
+        let mut interval = time::interval(time::Duration::from_millis(interval));
+
+        // Temporary structure to hold sortable values of `Event`.
+        struct SortableOutputEventsPayload {
+            batch: usize,
+            payload: OutputEventsPayload,
+        }
+
+        // Collect a vector of results, with a capacity of `limit`. As new `Event`s come in,
+        // they will be sampled and added to results.
+        let mut results = Vec::<SortableOutputEventsPayload>::with_capacity(limit);
+
+        // Random number generator to allow for sampling. Speed trumps cryptographic security here.
+        // The RNG must be Send + Sync to use with the `select!` loop below, hence `SmallRng`.
+        let mut rng = SmallRng::from_entropy();
```

Comment:
From the docs:

>  RNG recommended when small state, cheap initialization and good performance are required. The PRNG algorithm in SmallRng is chosen to be efficient on the current platform, without consideration for cryptography or security. The size of its state is much smaller than for StdRng

This is initialised once in an async task (each GraphQL subscription, and possibly each field, is a separate task), and is invoked a maximum of the total number of logs for the observed component(s) over a given interval, minus the provided `limit`.

Running the bench for [seq_iter_choose_from_1000](https://github.com/rust-random/rand/blob/671e0a29b0766ed629976b889c0fc7094e6fb1f0/benches/seq.rs#L77-L92) locally produces 5354 MB/s of usize values, which I think is ~701,759,488/sec on a 64 bit platform?

I don't think this is anything to be concerned with, but open to alternatives -- I chose this because it purports to be fast and thread safe.